slides/08_shiny.qmd

```diff
@@ -0,0 +1,442 @@
+This "stringy" reactivity graph is easier to understand and debug
+:::
+:::
+
+# Testing
+
+## Testing: From Icecream to Pyramid
+
+::: columns
+::: {.column width="50%"}
+Bad
+
+![](resources/icecream.png){height=400}
+
+- cannot always test manually
+- can easily introduce bugs
+:::
+ 
+::: {.column width="50%"}
+Good
+
+![](resources/pyramid.png){height=400}
+
+- manual testing is reduced
+- only for "playing around"
+:::
+:::
+
+## Testing: Snapshot tests for UI
+
+::: columns
+::: {.column width="50%"}
+```{r, echo = TRUE, eval = FALSE}
+#| code-line-numbers: "5-9"
+test_that("myInput UI works", {
+  input <- "foo"
+  set.seed(123)
+  datasets <- mock_datasets()
+  expect_snapshot(myInput(
+    "my_test",
+    datasets = datasets,
+    input = input
+  ))
+})
+```
+:::
+
+::: {.column width="50%"}
+- Shiny UI functions return shiny tag objects, i.e. HTML code
+- Can use snapshot tests to avoid accidental changes
+- Value is limited
+- Can create issues sometimes (non-reproducible hashes)
+:::
+:::
+
+## Testing: Shiny server tests
+
+::: columns
+::: {.column width="50%"}
+```{r, echo = TRUE, eval = FALSE}
+test_that("server works", {
+  testServer(server, {
+    session$setInputs(...)
+
+    print(reactive1())
+    print(output$output1)
+    # etc
+
+    # To interactively play:
+    # browser()
+  
+    expect_equal(...)
+    # etc.
+  })
+})
+
+```
+:::
+
+::: {.column width="50%"}
+- Code is run inside the server function
+- Can access reactives, outputs, etc.
+- Session object simulates user actions and time
+- Note: need to insert `browser()`
+- Also works for module server functions
+- Limitation: No UI, no JavaScript
+:::
+:::
+
+## Testing: Shiny app tests
+
+::: columns
+::: {.column width="50%"}
+```{r, echo = TRUE, eval = FALSE}
+test_that("my app works", {
+  library(shinytest)
+  app <- ShinyDriver$new(
+    "myAppDir/",
+    loadTimeout = 1e5, 
+    phantomTimeout = 1e5,
+    debug = "all"
+  ) # app$getDebugLog()
+  app$takeScreenshot()
+  app$setInputs(name = "Hadley")
+  app$getValue("greeting")
+  app$click("reset")
+  app$getValue("greeting")
+})
+```
+:::
+
+::: {.column width="50%"}
+- `{shinytest2}` is recent alternative
+- Usually the app is defined in `myAppDir/app.R`
+- New app instance from `ShinyDriver$new()`
+- Full Shiny app in a headless browser
+- Can look into it via `takeScreenshot()`
+- We can set inputs, get values of inputs/outputs, click buttons etc.
+:::
+:::
+
+# Additional Topics
+
+## Security
+
+- Never include passwords in your source code! Instead:
+  - Use environment variables
+  - Or use the `{config}` package
+  - Any files containing passwords must be added to `.gitignore`
+- For user authentication, start with existing solutions that are IT approved
+  - See some best practices [here](https://solutions.rstudio.com/sys-admin/auth/kerberos/) and [here](https://db.rstudio.com/best-practices/deployment/).
+- Code run in `server()` is isolated and cannot be seen by other users
```

Comment:
but careful about global env and options (because this could be shared by multiple sessions)